### PR TITLE
DIT - Permissions Refactor

### PR DESCRIPTION
**Pull request by ed0906**

### CODE REVIEW

File: app/uk/gov/hmrc/tariffclassificationfrontend/models/Permission.scala [hunk 1]
```diff
@@ -16,36 +16,273 @@
...
+
+  case object EDIT_SAMPLE extends CasePermission {
+    override def name: String = this.getClass.getName
+    override def appliesTo(`case`: Case, operator: Operator): Boolean = managersOrTeamMembersOnly(operator)
   }
 }
+
+
+
+//object Permission extends Enumeration {
```
Comment (ed0906) [inline app/uk/gov/hmrc/tariffclassificationfrontend/models/Permission.scala]: I'll remove this once I've fixed all the tests

File: app/uk/gov/hmrc/tariffclassificationfrontend/models/Permission.scala [hunk 2]
```diff
@@ -16,36 +16,232 @@
...
+  }
+
+  private def managersOrTeamMembersOnly(operator: Operator): Boolean = operator.role match {
+    case Role.CLASSIFICATION_MANAGER | Role.CLASSIFICATION_OFFICER => true
+    case _ => false
+  }
+
+  private def managersOnly(operator: Operator): Boolean = operator.role == Role.CLASSIFICATION_MANAGER
+
+  private def adminOnly(operator: Operator): Boolean = false
```
Comment (aaroncarroll82) [inline app/uk/gov/hmrc/tariffclassificationfrontend/models/Permission.scala]: This one still to be finalised?

Comment (ed0906) [reply app/uk/gov/hmrc/tariffclassificationfrontend/models/Permission.scala]: We don't have a role for it yet, so currently nobody has this

Comment (ed0906) [reply app/uk/gov/hmrc/tariffclassificationfrontend/models/Permission.scala]: Arguably those roles could be deleted I guess

Comment (aaroncarroll82) [reply app/uk/gov/hmrc/tariffclassificationfrontend/models/Permission.scala]: Personally I think we shouldn't leave stubs like this in if it's not implemented as it'll be confusing for a new team - if they have a task that involves adding a new type of role and supporting it, they'd add it all in at the time. What benefit do we give providing something unused at this stage?

Comment (ed0906) [reply app/uk/gov/hmrc/tariffclassificationfrontend/models/Permission.scala]: Agreed, I'll remove

**Comment by ed0906** [comment]
I'm going to pull out the case status logic in a separate PR

File: app/uk/gov/hmrc/tariffclassificationfrontend/controllers/AttachmentsController.scala [hunk 3]
```diff
@@ -110,7 +110,7 @@ class AttachmentsController @Inject()(verify: RequestActions,
   }
 
   def uploadAttachment(reference: String): Action[Either[MaxSizeExceeded, MultipartFormData[TemporaryFile]]] =
-    (verify.authenticated andThen verify.mustHave(Permission.ADD_ATTACHMENT))
+    (verify.authenticated andThen verify.casePermissions(reference) andThen verify.mustHave(Permission.ADD_ATTACHMENT))
```
Comment (SteveSugden) [inline app/uk/gov/hmrc/tariffclassificationfrontend/controllers/AttachmentsController.scala]: Why did you add verify.casePermissions?   Adding adding attachment is not dependant on case ownership

Comment (ed0906) [reply app/uk/gov/hmrc/tariffclassificationfrontend/controllers/AttachmentsController.scala]: ADD_ATTACHMENT is a `CasePermission`  and requires `verify.casePermissions`

Comment (ed0906) [reply app/uk/gov/hmrc/tariffclassificationfrontend/controllers/AttachmentsController.scala]: I could make  it a  `GlobalPermission` but that seems wrong

File: app/uk/gov/hmrc/tariffclassificationfrontend/models/Permission.scala [hunk 4]
```diff
@@ -16,36 +16,232 @@
...
+    override def name: String = nameOf(this)
+    override def appliesTo(`case`: Case, operator: Operator): Boolean = managersOrAssignedTeamMembersOnly(`case`, operator)
+  }
+
+  case object REMOVE_ATTACHMENTS extends CasePermission {
+    override def name: String = nameOf(this)
+    override def appliesTo(`case`: Case, operator: Operator): Boolean = managersOrAssignedTeamMembersOnly(`case`, operator)
+  }
+
+  case object CANCEL_CASE extends CasePermission {
```
Comment (SteveSugden) [inline app/uk/gov/hmrc/tariffclassificationfrontend/models/Permission.scala]: Should this not extend GlobalPermission?  Same with add-not, add-attachment, etc

Comment (ed0906) [reply app/uk/gov/hmrc/tariffclassificationfrontend/models/Permission.scala]: My thoughts were that `GlobalPermission`s  are for when you arent dealing with a single case e.g. Search, Queues etc

File: app/uk/gov/hmrc/tariffclassificationfrontend/controllers/AttachmentsController.scala (hunk 3 above)
Comment (SteveSugden) [reply app/uk/gov/hmrc/tariffclassificationfrontend/controllers/AttachmentsController.scala]: But it is a global permission - i.e. not dependent on ownership

File: app/uk/gov/hmrc/tariffclassificationfrontend/models/Permission.scala (hunk 4 above)
Comment (ed0906) [reply app/uk/gov/hmrc/tariffclassificationfrontend/models/Permission.scala]: This will also end up with a status check (`COMPLETED`) when I start pulling that out

Comment (SteveSugden) [reply app/uk/gov/hmrc/tariffclassificationfrontend/models/Permission.scala]: ... in which case you need the case (see what I did there).  Fair enough

File: app/uk/gov/hmrc/tariffclassificationfrontend/controllers/AttachmentsController.scala (hunk 3 above)
Comment (ed0906) [reply app/uk/gov/hmrc/tariffclassificationfrontend/controllers/AttachmentsController.scala]: `CasePermission`s  are specific to restricting things on a single `case`. I'd argue Adding attachments is specific to a single case

File: app/uk/gov/hmrc/tariffclassificationfrontend/models/Permission.scala [hunk 5]
```diff
@@ -16,36 +16,232 @@
...
+    override def name: String = nameOf(this)
+    override def appliesTo(`case`: Case, operator: Operator): Boolean = managersOrAssignedTeamMembersOnly(`case`, operator)
+  }
+
+  case object EDIT_LIABILITY extends CasePermission {
+    override def name: String = nameOf(this)
+    override def appliesTo(`case`: Case, operator: Operator): Boolean = managersOrAssignedTeamMembersOnly(`case`, operator)
+  }
+
+  case object EDIT_RULING extends CasePermission {
```
Comment (aaroncarroll82) [inline app/uk/gov/hmrc/tariffclassificationfrontend/models/Permission.scala]: Just wondering if you were planning to bring in the more complex permissions as part of this change, to make them part of the overall permissions trait? For instance the ones currently held in the LiabilityController along the lines of case is open and operator has permission for edit, **or**, case is new and operator has permission for create

Comment (SteveSugden) [reply app/uk/gov/hmrc/tariffclassificationfrontend/models/Permission.scala]: Stay tuned Aaron...  :-)

Comment (ed0906) [reply app/uk/gov/hmrc/tariffclassificationfrontend/models/Permission.scala]: Yep i'm going to do that as a second PR to keep them small

Comment (ed0906) [reply app/uk/gov/hmrc/tariffclassificationfrontend/models/Permission.scala]: Working on that next

Comment (aaroncarroll82) [reply app/uk/gov/hmrc/tariffclassificationfrontend/models/Permission.scala]: Sorry, the screen didn't refresh, I saw this afterwards -




> PR I'm going to pull out the case status logic in a separate

File: app/uk/gov/hmrc/tariffclassificationfrontend/models/Permission.scala [hunk 6]
```diff
@@ -16,36 +16,218 @@
...
+    EDIT_LIABILITY,
+    EDIT_RULING,
+    APPEAL_CASE,
+    EXTENDED_USE,
+    MOVE_CASE_BACK_TO_QUEUE,
+    EDIT_SAMPLE
+  )
+
+  def from(string: String): Option[Permission] = values.find(_.name == string)
+  def applyingTo(operator: Operator): Set[Permission] = values
```
Comment (pllaurado) [inline app/uk/gov/hmrc/tariffclassificationfrontend/models/Permission.scala]: All together is a great idea (Maybe more roles could mess but is not what this app is intended for)... not much to add here. Maybe I would define lazy structures to not iterate them again and again or some other little changes. But just a preference not a requirement 
`lazy val globalPermission =  values.filter(_.isInstanceOf[GlobalPermission]).map(_.asInstanceOf[GlobalPermission])`
`lazy val casePermissions = values.filter(_.isInstanceOf[CasePermission]).map(_.asInstanceOf[CasePermission])`
`def applyingTo(operator: Operator): Set[Permission] =  globalPermission.filter(_.appliesTo(operator)).map(_.asInstanceOf[Permission])`
